### DISCUSSION TITLE
Is there a way to rename the undefined1, undefined2, undefined4, undefined8 types?

### DISCUSSION THREAD

**Discussion by mrexodia** [NationalSecurityAgency/ghidra#2564] [Q&A] [answered]
Is there a way to rename these types?

![image](https://user-images.githubusercontent.com/2458265/102403271-4b1aaf00-3fe6-11eb-8e03-fd6a825ce25f.png)

I'd like them to be called `uint8_t`, `uint16_t` etc (or even IDA-style `BYTE`, `WORD`, `DWORD`).

**Comment by dev747368** [comment]
No, you can't rename built-in types.  You can create an alias / typedef of them with whatever name you want.

**Reply by mrexodia** [reply]
In that case, is there perhaps a way to make the decompiler automatically use my typedefs per default?

And thanks a lot for the quick answer!

**Reply by dev747368** [reply]
generally, the decompiler will use the data types that have been assigned in the listing, either the function parameters or the type of a static variable.

**Comment by emteere** [comment, marked as answer]
The undefined<x> datatypes are used to note an access of a size without knowing the real type, or the existence of a parameter without locking in the type.  They are really placeholder types.  The decompiler is free to propagate a type and replace the undefined with a good type for all undefined types.  In addition, undefined types defined in memory can be replaced with a more specific type by algorithms.

If the decompiler were to use uint8_t, that could be incorrect as the type might be signed.  It all depends on the access.  The decompiler will try to replace the type with a correct type based on the access or use.  Absent any other type information than a size, especially when accessing through an unknown pointer type, the decompiler will display undefined.  If the pointer points to a structure or class, the decompiler is more likely to use uint, char, byte, etc... as the type for local variables and parameters.

There is a setting to disable type casts, which could help improve default display if you don't specify types for parameters which you could try as well.

**Reply by mrexodia** [reply]
Thanks for the detailed explanation!
